### PR TITLE
[SYCL] Extend broadcast to TriviallyCopyable types

### PR DESCRIPTION
**Pull request by Pennycook**
Uses the TriviallyCopyable shuffle approach for broadcasts.

Note that this works for both work-groups and sub-groups, because
OpGroupBroadcast is defined as supporting both groups in SPIR-V.

Signed-off-by: John Pennycook <john.pennycook@intel.com>

---

Closes https://github.com/intel/llvm/issues/1885.  It's become clear while implementing this functionality that some refactoring of `spirv.hpp` and `group_algorithms.hpp` would be nice, since there is a lot of code duplication to handle cases that are essentially identical except for different types of cast.  I think such refactoring is beyond the scope of this PR, as the requirements of exactly which types should be supported and what casts are necessary is still evolving.

### CODE REVIEW

File: sycl/include/CL/sycl/detail/spirv.hpp [hunk 1]
```diff
@@ -41,47 +67,133 @@ template <typename Group> bool GroupAny(bool pred) {
...
+
+template <typename T>
+using ConvertToNativeBroadcastType_t = select_cl_scalar_integral_unsigned_t<T>;
+
+// Generic broadcasts may require multiple calls to SPIR-V GroupBroadcast
+// intrinsics, and should use the fewest broadcasts possible
+// - Loop over 64-bit chunks until remaining bytes < 64-bit
+// - At most one 32-bit, 16-bit and 8-bit chunk left over
+template <typename T, typename IdT = size_t>
+using EnableIfGenericBroadcast =
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/detail/spirv.hpp:94]: I do not see difference between this and EnableIfBitcastBroadcast. is it expected?

Comment (Pennycook) [reply sycl/include/CL/sycl/detail/spirv.hpp:94]: There's an extra `!` in front of the second condition, which I agree is hard to see.  I'll have a think and see if there's a simpler way to express these conditions.

Comment (Pennycook) [reply sycl/include/CL/sycl/detail/spirv.hpp]: I've tried to tidy things up a little bit in 0bdb196.  Is it clearer?

**Comment by Pennycook** [comment]
Resolved merge conflicts.

**Comment by Pennycook** [comment]
@vladimirlaz, @AlexeySachkov: I think this one is ready for review now.

File: sycl/include/CL/sycl/detail/spirv.hpp [hunk 2]
```diff
@@ -41,47 +67,137 @@ template <typename Group> bool GroupAny(bool pred) {
...
+  T Result;
+  char *XBytes = reinterpret_cast<char *>(&x);
+  char *ResultBytes = reinterpret_cast<char *>(&Result);
+  auto BroadcastBytes = [=](size_t Offset, size_t Size) {
+    uint64_t BroadcastX, BroadcastResult;
+    detail::memcpy(&BroadcastX, XBytes + Offset, Size);
+    BroadcastResult = GroupBroadcast<Group>(BroadcastX, local_id);
+    detail::memcpy(ResultBytes + Offset, &BroadcastResult, Size);
+  };
+  GenericShuffle<T>(BroadcastBytes);
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/detail/spirv.hpp:146]: It looks a bit strange to see a broadcast implemented via shuffle. Probably it makes sense to rename `GenericShuffle` to something like "ForEach..something?". I can't come up with a name right now, probably you have any ideas? I wanted to express the idea that functor is applied for each "native" chunk of data in `T`

Comment (Pennycook) [reply sycl/include/CL/sycl/detail/spirv.hpp]: I agree.  This is one of the things that I lazily lumped under the grand refactoring TODO. :)

I have a suspicion that something like this function is going to be useful for implementing other generic versions of SPIR-V functions, but I'm also struggling to come up with a better way to represent it.

I've implemented something in f0a1e2f2a697dadc54bb0539cd2cfa3baad1072b.  It's still not perfect, but is hopefully clearer than what was there before, and should help anybody trying to make sense of the code before these header files get tidied up.

File: sycl/include/CL/sycl/detail/spirv.hpp (hunk 1 above)
Comment (vladimirlaz) [reply sycl/include/CL/sycl/detail/spirv.hpp]: yes. thank you

**Pull request merged**